File: problem10/cpp-solution/p10.cpp
/*
 * p10.cpp
 *
 * Zeke Reyna
 *
 * Problem 10:
 * Find the sum of all the primes below two million.
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
using ull = unsigned long long;
const int _BOUND_DEF = 2000000;

vector<int> generate_primes(int bound)
{
	vector<int> primes;
	for (int num = 2; num < bound; num++) {
		bool isPrime = none_of(primes.begin(), primes.end(),
				[num](const int p) { return num % p == 0; });
		if (isPrime)
			primes.push_back(num);
	}
	return primes;
}

ull sum_primes(vector<int>& primes)
{
	return accumulate(primes.begin(), primes.end(), 0);
}

int main(int argc, char **argv)
{
	ios_base::sync_with_stdio(false);
	int n;
	(argc == 2) ? (n = stoi(argv[1])) : (n = _BOUND_DEF);

	vector<int> primes = generate_primes(n);
	ull sum = sum_primes(primes);

	cout << "\nThe sum of the primes below " << n;
	cout << " is " << sum << "\n\n";
	return 0;
}

File: problem10/java-solution/P10.java
/*
 * P10.java
 *
 * Zeke Reyna
 *
 * Problem 10:
 * Find the sum of all the primes below two million.
 */

import java.util.ArrayDeque;
import java.util.Queue;
import java.lang.Math.*;

public class P10
{
	public static Queue<Integer> generate_primes(int bound)
	{
		Queue<Integer> primes = new ArrayDeque<Integer>();
		for (int num = 2; num < bound; num++) {
			boolean isPrime = true;
			for (Integer p : primes) {
				if (num % p == 0) {
					isPrime = false;
					break;
				}
			}
			if (isPrime)
				primes.add(num);
		}
		return primes;
	}

	public static long sum_primes(Queue<Integer> primes)
	{
		long sum = 0;
		for (Integer num : primes)
			sum += num;
		return sum;
	}

	public static void main(String args[])
	{
		int n;
		long sum;
		Queue<Integer> primes;

		n = (args.length == 1) ? Integer.parseInt(args[0]) : 2000000;

		primes = generate_primes(n);
		sum = sum_primes(primes);

		System.out.print("\nThe sum of the primes below ");
		System.out.print(n + " is " + sum + "\n\n");
	}
}
